[Crystal] iP

crvstalphua opened the pull request:
## Bob - Remembering every bit and bob
> "Take care of all your memories. For you cannot relive them." - Bob Dylan ([source](https://www.brainyquote.com/quotes/bob_dylan_142081))

### What is a Bob?
Bob helps you remember, it's
- text-based
- _easy_ to use
- ~FRIENDLY~ rEalLY FrIENDly 
- best of all... it's **FREE!**

### I want BOB!  🤩
All you need to do is:
1. download it [here](https://github.com/crvstalphua/ip/releases/download/A-Jar/Bob.jar)
2. in the terminal, navigate to the folder containing file and run `java -jar Bob.jar`
3. AND.... done! start using Bob to store your tasks! 🤭

### What can Bob do?

Here's what Bob can do:
- [X] Add/Remove a task
- [X] Note if task is completed or not
- [X] List down all current tasks
- [x] Filter all tasks of the same deadline
- [ ] Set Reminders (SOON!)

### Just a snippet into Bob's main method 
```java
public void run() {
        ui.displayWelcomeMessage();
        while (parser.getToExit() == false) {
            parser.parse(ui.getReply(), taskList, storage, ui);
        }
    }

    public static void main(String[] args) {
        new Bob().run();
    }
```

## Bob remembers things for you, so you don't have to!

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

Commit:
added *.class to .gitignore and created skeletal version of Bob

Commit:
added list for tasks

Commit:
added task.java and feature for marking tasks as done

Commit:
added ToDo, Deadline, Event subclasses

Commit:
added automated text ui testing

Commit:
Added DukeException class (not in use) and handled errors due to incorrect input

Commit:
added remove feature

Commit:
added feature to save tasks, and keep track of tasks when Bob is restarted

Commit:
edited file reader to also read if task is marked

Commit:
added feature to read dates of a yyyy-mm-dd format and method to filter only tasks of a specific date

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  added feature to read dates of a yyyy-mm-dd format and method to filter only tasks of a specific date

# Conflicts:
#	src/main/java/Bob.java
#	src/main/java/Deadline.java
#	src/main/java/Event.java

Commit:
added parser, storage, ui and tasklist classes.
added bob package

Commit:
edited and simplified textfile reading in storage class,

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'
add gradle support

Commit:
successfully added gradle

Commit:
added junit tests for event, deadline and todo classes

Commit:
modified build.gradle to allow for jar file saving

Commit:
added javadoc headers for all classes and methods

Commit:
Parser class: add getToExit() method to retrieve state of toExit boolean

Edit code to satisfy Coding Standard

Commit:
Add feature to search for tasks with specific keyword

Ui class: Modify displayTaskList() to take in taskList and message, and display a list of tasks with the given message
Enable method to be used to print a list, given any taskList (e.g. filtered taskList)

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Parser class: add getToExit() method to retrieve state of toExit boolean

# Conflicts:
#	src/main/java/bob/Parser.java
#	src/main/java/bob/Storage.java
#	src/main/java/bob/Ui.java

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  added javadoc headers for all classes and methods

# Conflicts:
#	src/main/java/bob/Storage.java
#	src/main/java/bob/TaskList.java
#	src/main/java/bob/Ui.java

Commit:
Parser class: Add header for getToExit()

shaniceng commented on src/main/java/bob/Task.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @return "[X]" if task is done and "[ ]" if task is not
+     */
+    public String getStatusIcon() {
```
Great job in naming all the methods in verbs and in camel cases 💯

shaniceng commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> tasks;
```
Great job in naming array list as plural 👍

shaniceng commented on src/main/java/bob/Event.java:
```diff
@@ -0,0 +1,41 @@
+
+    protected LocalDate at;
+
+    /**
```
Good job in commenting all of your classes as well as methods clearly 💯

shaniceng commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,153 @@
+    public void parse(String reply, TaskList tasks, Storage storage, Ui ui) {
+        String[] splitReply = reply.split(" ");
+        String command = splitReply[0].toLowerCase();
+        switch(command) {
```
Good job in using the correct indentations for a switch statement 👍

shaniceng commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,153 @@
+            ui.printLine();
+            this.toExit = true;
+            break;
+        case "lift":
```
Should this be "list" instead of "lift"?
```suggestion
        case "list":
```

crvstalphua replied:
Yes, almost missed that, thank you!

shaniceng reviewed (commented):
Overall, you did a really great job in all of the naming and coding standard 👍 
I found it easy to read is most cases and it LGTM, just some nits that I have suggested if you would like to change :)

leecaregene commented on src/main/java/bob/Task.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param x boolean value indicating if task should be marked as completed or not
+     */
+    public void toMark(boolean x) {
```
Perhaps you can rename this function to `mark` to make its purpose clearer?
```suggestion
    public void mark(boolean x) {
```

crvstalphua replied:
Okay, thank you!

leecaregene commented on src/main/java/bob/Task.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @return String representing how Task object is saved
+     */
+    public String toSave() {
```
Perhaps you can rename this function to `toSaveFormat` to make its purpose clearer?
```suggestion
    public String toSaveFormat() {
```

crvstalphua replied:
Thank you, will edit it!

leecaregene commented on src/test/java/bob/DeadlineTest.java:
```diff
@@ -0,0 +1,22 @@
+
+public class DeadlineTest {
+    @Test
+    public void testOne() {
```
Perhaps you can make the names of the test functions more descriptive?

crvstalphua replied:
Thanks for the suggestion! Will edit the test names!

leecaregene commented on src/main/java/bob/Ui.java:
```diff
@@ -0,0 +1,128 @@
+     * @param tasks list of tasks
+     * @param index index of unmarked task
+     */
+    public void displayUnMarked(TaskList tasks, int index) {
```
`displayUnmarked` looks nicer in my opinion 😛
```suggestion
    public void displayUnmarked(TaskList tasks, int index) {
```

crvstalphua replied:
Haha, you're right :")

crvstalphua committed:
Update src/main/java/bob/Parser.java

Co-authored-by: shaniceng <57250131+shaniceng@users.noreply.github.com>

Commit:
Task/Deadline/ToDo/Event: change toSave to toSaveFormat
Task: change toMark to mark
Ui: change displayUnMarked to displayUnmarked

Commit:
Merge commit '298b868c2d73853dc4deffee1e26fe4d32b55593'

* commit '298b868c2d73853dc4deffee1e26fe4d32b55593':
  Update src/main/java/bob/Parser.java

Commit:
Add checkstyle

Commit:
Add Gui for Bob

Commit:
GUI: fix issue of dialog box cutting off text when it is too long, decrease image size, add padding around image

Commit:
Ui: Add assertions for methods
TaskList: Add assertion for getLength() method
Parser: Add catch block for "mark", "unmark" and "remove" cases

For methods in Ui and TaskList, the length of the task list cannot be negative.
For parser, error is thrown when index is out of bounds.

The task list should never be negative, otherwise it indicates an error in the program.
It is more user-friendly to remind the user that the index is out of bounds rather than to terminate the entire program.

Let's
* add assertions in the various methods of Ui and getLength() in TaskList to assert that length of task list has to be positive
* add catch block for "mark", "unmark" and "remove" cases of Parser, to print an error message for the user when index is out of bounds

Commit:
Improve Code Quality

Some classes and methods lack javaDoc headers. In particular, code for the different enums in Parser class are deeply nested and hard to read and understand.

Fixing the code quality makes the code easier to read and understand for users or other future developers who may reference the code.

Let's
* add javadoc comments and headers for all class and methods
* make sure code is not too long
* create new Command classes to remove deep nesting in Parser class

crvstalphua committed:
Merge pull request #3 from crvstalphua/branch-A-CodeQuality

Improve Code Quality

Commit:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Ui: Add assertions for methods TaskList: Add assertion for getLength() method Parser: Add catch block for "mark", "unmark" and "remove" cases

# Conflicts:
#	src/main/java/bob/Parser.java
#	src/main/java/bob/Ui.java

Commit:
Add Update Command
Name and date of tasks cannot be editted without deleting the task and adding a new one.

Adding an update feature allows users to edit tasks without having to delete them, making it more user-friendly and convenient. It also prevents the changing of task indexs in the list, if the user arranged it intentionally.

Let's
* add a update command class
* add update methods in each task class

Commit:
Add Help Command

By entering "help", bob will reply with the full list of commands.

Commit:
Edit README.md

Commit:
Fix minor issues in README.md

Commit:
Add Ui.png

Commit:
Add javadoc header for displayCommands() method in Ui

Commit:
Fix a few minor issues
*Add test cases for Parser class, DeadlineCommand class and EventCommand Class
*Fix auto-resize and scaling issue of GUI
*Fix checkstyle issues
